python/GafferSceneTest/IECoreScenePreviewTest/PrimitiveAlgoTest.py

```diff
@@ -0,0 +1,537 @@
+
+		file = IECoreScene.SceneInterface.create(
+			str( self.usdFileDir / "generalTestMesh.usd" ), IECore.IndexedIO.OpenMode.Read
+		)
+		meshSource = file.child( "object" ).readObject( 0.0 )
+		for k in meshSource.keys():
+			if k == "P":
+				continue
+			del meshSource[k]
+
+		random.seed( 42 )
+
+		def randomC():
+			return imath.Color3f( random.random(), random.random(), random.random() )
+
+		curvesSource = IECoreScene.CurvesPrimitive(
+			IECore.IntVectorData( [ 4, 5, 6, 5, 4] ), IECore.CubicBasisf.linear(), False
+		)
+		curvesSourceCubic = IECoreScene.CurvesPrimitive(
+			IECore.IntVectorData( [ 4, 5, 6, 5, 4] ), IECore.CubicBasisf.bSpline(), False
+		)
+		curvesSourcePeriodic = IECoreScene.CurvesPrimitive(
+			IECore.IntVectorData( [ 4, 5, 6, 5, 4] ), IECore.CubicBasisf.linear(), True
+		)
+		curvesSourcePeriodicCubic = IECoreScene.CurvesPrimitive(
+			IECore.IntVectorData( [ 4, 5, 6, 5, 4] ), IECore.CubicBasisf.bSpline(), True
+		)
+
+		for c in [ curvesSource, curvesSourceCubic, curvesSourcePeriodic, curvesSourcePeriodicCubic ]:
+			c["P"] = IECoreScene.PrimitiveVariable( Interpolation.Vertex,
+				IECore.V3fVectorData( [ imath.V3f( randomC() ) for i in range( c.variableSize( Interpolation.Vertex ) ) ] )
+			)
+
+		pointsSource = IECoreScene.PointsPrimitive( 20 )
+		pointsSource["P"] = IECoreScene.PrimitiveVariable( Interpolation.Vertex,
+			IECore.V3fVectorData( [ imath.V3f( randomC() ) for i in range( 20 ) ] )
+		)
+
+		for source in [ meshSource, curvesSource, curvesSourceCubic, curvesSourcePeriodic, curvesSourcePeriodicCubic, pointsSource ]:
+			isMesh = type( source ) == IECoreScene.MeshPrimitive
+			isCurves = type( source ) == IECoreScene.CurvesPrimitive
+			isPoints = not isMesh and not isCurves
+
+			# Create a source with one primvar of each interpolation
+			allInterps = source.copy()
+			allInterps["A"] = IECoreScene.PrimitiveVariable( Interpolation.Constant, IECore.Color3fData( randomC() ) )
+			for name, interp in [
+				( "B", Interpolation.Vertex ), ( "C", Interpolation.Uniform ),
+				( "D", Interpolation.Varying ), ( "E", Interpolation.FaceVarying )
+			]:
+				allInterps[name] = IECoreScene.PrimitiveVariable(
+					interp, IECore.Color3fVectorData( [ randomC() for i in range( source.variableSize( interp ) ) ] )
+				)
+			allInterps["labelSource"] = IECoreScene.PrimitiveVariable( Interpolation.Constant, IECore.IntData( 0 ) )
+
+			for interp, indexed in itertools.product( [
+					Interpolation.Constant, Interpolation.Vertex, Interpolation.Uniform,
+					Interpolation.Varying, Interpolation.FaceVarying
+				], [ False, True ] ):
+
+				# Create a source with all primvars set to the current interpolation, optionally indexed
+				oneInterp = source.copy()
+				for name in [ "A", "B", "C", "D", "E" ]:
+					varSize = source.variableSize( interp )
+
+					if interp == Interpolation.Constant:
+						oneInterp[name] = IECoreScene.PrimitiveVariable( interp, IECore.Color3fData( randomC() ) )
```

Comment:
There's already a basic test here:
https://github.com/GafferHQ/gaffer/blob/334a7b2d3d61086454666821e215cd93f0cbea96/python/GafferSceneTest/IECoreScenePreviewTest/PrimitiveAlgoTest.py#L545

I'm assuming that in order to be indexed, a Constant would have to be a vector type, so it would always just be discarded, because we merge Constants into Uniforms, and we don't support vectors of vectors.

Reply:
Good point - thanks for the explanation.